File: uishell/src/main/scala/net/kemitix/thorp/uishell/UploadEventListener.scala
package net.kemitix.thorp.uishell

import java.util.concurrent.atomic.AtomicLong

import net.kemitix.thorp.domain.{Channel, LocalFile}
import net.kemitix.thorp.uishell.UploadProgressEvent.RequestEvent

object UploadEventListener {

  final case class Settings(uiSink: Channel.Sink[UIEvent],
                            localFile: LocalFile,
                            index: Int,
                            totalBytesSoFar: Long,
                            batchMode: Boolean)

  def listener(settings: Settings): UploadProgressEvent => Unit = {
    val bytesTransferred = new AtomicLong(0L)
    event =>
      {
        event match {
          case e: RequestEvent =>
            settings.uiSink.accept(
              UIEvent.requestCycle(
                settings.localFile,
                bytesTransferred.addAndGet(e.transferred),
                settings.index,
                settings.totalBytesSoFar
              )
            )
          case _ => ()
        }
      }
  }

}

File: uishell/src/main/java/net/kemitix/thorp/uishell/UploadEventListener.java
package net.kemitix.thorp.uishell;

import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import net.kemitix.thorp.domain.Channel;
import net.kemitix.thorp.domain.LocalFile;
import net.kemitix.thorp.uishell.UploadProgressEvent.RequestEvent;

import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Consumer;

public class UploadEventListener {

    Consumer<UploadProgressEvent> listener(Settings settings) {
        AtomicLong bytesTransferred = new AtomicLong(0L);
        return event -> {
            if (event instanceof RequestEvent) {
                RequestEvent requestEvent = (RequestEvent) event;
                settings.uiSink.accept(
                        UIEvent.requestCycle(
                                settings.localFile,
                                bytesTransferred.addAndGet(requestEvent.transferred),
                                settings.index,
                                settings.totalBytesSoFar
                        )
                );
            }
        };
    }

    public static Settings settings(
            Channel.Sink<UIEvent> uiSink,
            LocalFile localFile,
            int index,
            long totalBytesToFar,
            boolean batchMode
    ) {
        return new Settings(uiSink, localFile, index, totalBytesToFar);
    }
    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
    public static class Settings {
        public final Channel.Sink<UIEvent> uiSink;
        public final LocalFile localFile;
        public final int index;
        public final long totalBytesSoFar;
    }
}
